### PR TITLE
WIP: Generate Erlang test coverage report

### PR DESCRIPTION
**Pull request by lrascao**
This is mostly a WIP/PoC and can be dropped if not desired. I figured it would be useful to guide adding/tweaking of test cases, of course this must be the most hackish possible way of achieving it, if we think it's useful then we can clean it up.
Here is the coverage output:

```
Cover data written to "./elixir.coverdata".
Performing cover analysis...  |------------------------|------------|
  |                module  |  coverage  |
  |------------------------|------------|
  |                elixir  |       66%  |
  |        elixir_aliases  |       95%  |
  |      elixir_bitstring  |       94%  |
  |      elixir_bootstrap  |       50%  |
  |        elixir_clauses  |       96%  |
  |    elixir_code_server  |       76%  |
  |       elixir_compiler  |       65%  |
  |         elixir_config  |       66%  |
  |            elixir_def  |       94%  |
  |       elixir_dispatch  |       94%  |
  |            elixir_env  |      100%  |
  |            elixir_erl  |       91%  |
  |   elixir_erl_compiler  |       59%  |
  |        elixir_erl_for  |      100%  |
  |       elixir_erl_pass  |       98%  |
  |        elixir_erl_var  |       87%  |
  |         elixir_errors  |       89%  |
  |         elixir_expand  |       95%  |
  |             elixir_fn  |       98%  |
  |         elixir_import  |       79%  |
  |  elixir_interpolation  |       80%  |
  |        elixir_lexical  |       92%  |
  |         elixir_locals  |       91%  |
  |            elixir_map  |       94%  |
  |         elixir_module  |       90%  |
  |    elixir_overridable  |      100%  |
  |         elixir_parser  |       34%  |
  |          elixir_quote  |       88%  |
  |        elixir_rewrite  |       49%  |
  |            elixir_sup  |        0%  |
  |      elixir_tokenizer  |       85%  |
  |          elixir_utils  |       79%  |
  |------------------------|------------|
  |                 total  |       48%  |
  |------------------------|------------|
  coverage calculated from:
    ./elixir.coverdata
  cover summary written to: /Users/luis/Projects/elixir/lib/elixir/index.html
Finished up in 91.1 seconds (74.9s on load, 9.0s async, 7.2s sync)
1729 doctests, 3678 tests, 0 failures, 9 excluded, 16 invalid

Randomized with seed 707732
```

### CODE REVIEW

File: lib/elixir/test/erlang/test_helper.erl [hunk 1]
```diff
@@ -35,3 +41,247 @@ throw_erlang(String) ->
   {ok, Tokens, _} = erl_scan:string(String),
   {ok, [Form]} = erl_parse:parse_exprs(Tokens),
   erlang:error(io:format("~p~n", [Form])).
+
+cover_compile() ->
```
Comment (lrascao) [inline lib/elixir/test/erlang/test_helper.erl:45]: everything below is a shameless rip off of rebar3's `rebar_prv_cover.erl`

File: lib/ex_unit/lib/ex_unit/formatter.ex [hunk 2]
```diff
@@ -110,7 +110,10 @@ defmodule ExUnit.Formatter do
         ""
       end
 
-    "Finished in #{format_us(total_us)} seconds " <>
+    :test_helper.maybe_write_coverdata()
```
Comment (lrascao) [inline lib/ex_unit/lib/ex_unit/formatter.ex:113]: this is the easiest way i could find of generating the coverage report after running all the tests

File: Makefile [hunk 3]
```diff
@@ -286,7 +286,7 @@ test_stdlib: compile
 	$(Q) if [ "$(OS)" = "Windows_NT" ]; then \
 		cd lib/elixir && cmd //C call ../../bin/elixir.bat -r "test/elixir/test_helper.exs" -pr "test/elixir/**/$(TEST_FILES)"; \
 	else \
-		cd lib/elixir && ../../bin/elixir -r "test/elixir/test_helper.exs" -pr "test/elixir/**/$(TEST_FILES)"; \
+		cd lib/elixir && ../../bin/elixir -r "test/elixir/test_helper.exs" -r "test/elixir/**/$(TEST_FILES)" -pa "test/ebin"; \
```
Comment (lrascao) [inline Makefile:289]: allows us to call into `test_helper`

File: lib/elixir/test/elixir/test_helper.exs [hunk 4]
```diff
@@ -87,6 +87,8 @@ assert_timeout = String.to_integer(System.get_env("ELIXIR_ASSERT_TIMEOUT") || "5
 epmd_exclude = if match?({:win32, _}, :os.type()), do: [epmd: true], else: []
 os_exclude = if PathHelpers.windows?(), do: [unix: true], else: [windows: true]
 
+:test_helper.cover_compile()
```
Comment (lrascao) [inline lib/elixir/test/elixir/test_helper.exs:90]: almost all (except two that crash on instrumentation) erlang modules are cover compiled

**Comment by josevalim** [comment]
Hi @lrascao, thank you for the PR. This is definitely a lot of code for a functionality that we likely won't use much, so we won't be moving forward. Thanks!

PS: There is also a good amount of false positives in the coverage because some of the code runs during boot/compilation. :)

**Pull request closed without merging**